examples/neighbors/pamm.py

```diff
@@ -0,0 +1,405 @@
+
+    fig, ax = plt.subplots(subplot_kw={"projection": "3d"}, figsize=fig_size, dpi=100)
+    scatter = ax.scatter(
+        grids[:, use_index[0]],
+        grids[:, use_index[1]],
+        grids[:, use_index[2]],
+        c=grid_label_,
+        s=grid_weights * size_scale,
+    )
+    legend1 = ax.legend(*scatter.legend_elements(), loc="lower left", title="Gaussian")
+    ax.add_artist(legend1)
+    ax.set_xlabel(label_text[0])
+    ax.set_ylabel(label_text[1])
+    ax.set_zlabel(label_text[2])
+
+    return fig, ax
+
+
+# %%
+# We first load our dataset:
+#
+#
+
+# %%
+hbond_data = load_hbond_dataset()
+descriptors = hbond_data["descriptors"]
+weights = hbond_data["weights"]
+
+# %%
+# We use the `FPS` class to select the `ngrid` descriptors with the highest. It is
+# recommended to set the number of grids as the square root of the number of
+# descriptors. Then we do the fit of the KDE.
+
+
+# %%
+ngrid = int(len(descriptors) ** 0.5)
+selector = FPS(initialize=26310, n_to_select=ngrid)
+selector.fit(descriptors.T)
+selector.selected_idx_
+grids = descriptors[selector.selected_idx_]
+
+# %%
+estimator = SparseKDE(descriptors, weights)
+estimator.fit(grids)
+
+# %%
+# Now we visualize the distribution and the weight of clusters.
+
+# %%
+cluster_distribution_3D(
+    grids, estimator._sample_weights, label_text=[r"$\nu$", r"$\mu$", r"r"]
+)
+
+# %%
+# We need to estimate the probability at each grid point to do quick shift, which can
+# further partition the set of grid points in to several clusters. The resulting
+# clusters can be interpreted as (meta-)stable states of the system.
+#
+#
+
+# %%
+probs = estimator.score_samples(grids)
+qscuts = np.array([np.trace(cov) for cov in estimator._covariance])
+clustering = QuickShift(
+    qscuts**2,
+    metric_params=estimator.metric_params,
+)
+clustering.fit(grids, samples_weight=probs)
+cluster_centers_idx = clustering.cluster_centers_idx_
+labels = clustering.labels_
+normpks = logsumexp(probs)
+
+cluster_centers, labels = quick_shift_refinement(
+    grids,
+    cluster_centers_idx,
+    labels,
+    probs,
+    estimator.metric,
+    estimator.cell,
+)
+
+# %%
+# Based on the results, the gaussian mixture model of the system can be generated:
+#
+#
+
+# %%
+cluster_weights, cluster_means, cluster_covs, labels = generate_probability_model(
+    cluster_centers_idx,
+    labels,
+    grids,
+    estimator.descriptors,
+    estimator._sample_labels_,
+    estimator.weights,
+    probs,
+    estimator.cell,
+)
+
+# %%
+# The final result shows seven (meta-)stable states of hydrogen bond. Here we also show
+# the reference hydrogen bond descriptor. The gaussian with the largest weight locates
+# closest to the reference point. This result shows that, with the help of the
+# `SparseKDE` and `QuickShift` algorithm, we can easily identify the (meta-)stable
+# states of the system objectively and without any prior knowledge about the system.
+#
+#
+
+# %%
+REF_HB = np.array([0.82, 2.82, 2.74])
```

Comment:
Can you explain this variable and why you mark it in the plot?

Reply:
This is the coordinate of the "standard" hydrogen bond. I will add some explanation for it.